pydantic/networks.py

```diff
@@ -79,13 +79,110 @@ def __hash__(self) -> int:
+- `path`: optional path (`/the/path/`).
+- `query`: optional URL query (for example, `GET` arguments or "search string", such as `query=here`).
+- `fragment`: optional fragment (`fragment=is;this=bit`).
+"""
 AnyHttpUrl = Annotated[Url, UrlConstraints(allowed_schemes=['http', 'https'])]
-"""A type that will accept any http or https URL."""
+"""A type that will accept any http or https URL.
+
+* TLD not required
+* Host required
+"""
 HttpUrl = Annotated[Url, UrlConstraints(max_length=2083, allowed_schemes=['http', 'https'])]
-"""A type that will accept any http or https URL with a max length of 2083 characters."""
+"""A type that will accept any http or https URL.
+
+* TLD required
+* Host required
+* Max length 2083
+
+```py
+from pydantic import BaseModel, HttpUrl, ValidationError
+
+class MyModel(BaseModel):
+    url: HttpUrl
+
+m = MyModel(url='http://www.example.com')
+print(m.url)
+#> http://www.example.com/
+
+try:
+    MyModel(url='ftp://invalid.url')
+except ValidationError as e:
+    print(e)
+    '''
+    1 validation error for MyModel
+    url
+      URL scheme should be 'http' or 'https' [type=url_scheme, input_value='ftp://invalid.url', input_type=str]
+    '''
+
+try:
+    MyModel(url='not a url')
+except ValidationError as e:
+    print(e)
+    '''
+    1 validation error for MyModel
+    url
+      Input should be a valid URL, relative URL without a base [type=url_parsing, input_value='not a url', input_type=str]
+    '''
+```
+
+"International domains" (e.g. a URL where the host or TLD includes non-ascii characters) will be encoded via
+[punycode](https://en.wikipedia.org/wiki/Punycode) (see
+[this article](https://www.xudongz.com/blog/2017/idn-phishing/) for a good description of why this is important):
+
+```py
+from pydantic import BaseModel, HttpUrl
+
+class MyModel(BaseModel):
+    url: HttpUrl
+
+m1 = MyModel(url='http://puny£code.com')
+print(m1.url)
+#> http://xn--punycode-eja.com/
+m2 = MyModel(url='https://www.аррӏе.com/')
+print(m2.url)
+#> https://www.xn--80ak6aa92e.com/
+m3 = MyModel(url='https://www.example.珠宝/')
+print(m3.url)
+#> https://www.example.xn--pbt977c/
+```
+
+
+!!! warning "Underscores in Hostnames"
+    In Pydantic, underscores are allowed in all parts of a domain except the TLD.
+    Technically this might be wrong - in theory the hostname cannot have underscores, but subdomains can.
+
+    To explain this; consider the following two cases:
+
+    - `exam_ple.co.uk`: the hostname is `exam_ple`, which should not be allowed since it contains an underscore.
+    - `foo_bar.example.com` the hostname is `example`, which should be allowed since the underscore is in the subdomain.
+
+    Without having an exhaustive list of TLDs, it would be impossible to differentiate between these two. Therefore
+    underscores are allowed, but you can always do further validation in a validator if desired.
+
+    Also, Chrome, Firefox, and Safari all currently accept `http://exam_ple.com` as a URL, so we're in good
```

Comment:
Is it?
```suggestion
    Also, Chrome, Firefox, and Safari all currently accept `http://example.com` as a URL, so we're in good
```

Reply:
Yes, from https://github.com/pydantic/pydantic/pull/7497/files#diff-a09e6ceb30be505de26d94637fdd2127c78bc2fefa6e0de4eface9e1a1e20d2eL192